plugins/techdocs-backend/src/search/DefaultTechDocsCollator.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PluginEndpointDiscovery } from '@backstage/backend-common';
+import { Entity, RELATION_OWNED_BY } from '@backstage/catalog-model';
+import { IndexableDocument, DocumentCollator } from '@backstage/search-common';
+import fetch from 'cross-fetch';
+import unescape from 'lodash/unescape';
+import { Logger } from 'winston';
+import pLimit from 'p-limit';
+import { CatalogApi, CatalogClient } from '@backstage/catalog-client';
+
+interface MkSearchIndexDoc {
+  title: string;
+  text: string;
+  location: string;
+}
+
+export interface TechDocsDocument extends IndexableDocument {
+  kind: string;
+  namespace: string;
+  name: string;
+  lifecycle: string;
+  owner: string;
+}
+
+export class DefaultTechDocsCollator implements DocumentCollator {
+  protected discovery: PluginEndpointDiscovery;
+  protected locationTemplate: string;
+  private readonly logger: Logger;
+  private readonly catalogClient: CatalogApi;
+  private readonly parallelismLimit: number;
+  public readonly type: string = 'techdocs';
+
+  constructor({
+    discovery,
+    locationTemplate,
+    logger,
+    catalogClient,
+    parallelismLimit = 10,
+  }: {
+    discovery: PluginEndpointDiscovery;
+    logger: Logger;
+    locationTemplate?: string;
+    catalogClient?: CatalogApi;
+    parallelismLimit?: number;
+  }) {
+    this.discovery = discovery;
+    this.locationTemplate =
+      locationTemplate || '/docs/:namespace/:kind/:name/:path';
+    this.logger = logger;
+    this.catalogClient =
+      catalogClient || new CatalogClient({ discoveryApi: discovery });
+    this.parallelismLimit = parallelismLimit;
+  }
+
+  async execute() {
+    const limit = pLimit(this.parallelismLimit);
+    const techDocsBaseUrl = await this.discovery.getBaseUrl('techdocs');
+    const entities = await this.catalogClient.getEntities({
```

Comment:
I'm wondering if this will need to be paginated eventually, thinking that maybe a full dump of all the entities could be pretty heavy on both the catalog and the client in terms of memory, but maybe we can test it and see how far we get.

Reply:
Yup, good point. I have a feeling we can safely go into thousands before we might start seeing some effects due to the payload size (depending on the setup of course). In the end this data is only temporarily kept in memory and the subsequent heavier call to get the actual docs is throttled.

Reply:
Yeah I thought maybe that the pagination could drive the refresh from the docs too, so for each page of the catalog go refresh that, and then start on the next page. It spreads out the load across both the services that way, and you get throttling built in that way rather than trying to do all the things at once :)

Reply:
Yeah makes sense. I think the whole purge-all-stale -> index-all is a bit cumbersome in general, especially with large amounts of entries. Might get tricky to handle stale data invalidation etc. though if slicing updates to partial indexes.